connect/src/receive/connect.rs

```diff
@@ -0,0 +1,79 @@
+use crate::ConnectError;
+
+use timely::dataflow::operators::capture::EventReader;
+
+use std::sync::{Arc, Mutex};
+use std::net::{TcpStream, TcpListener, ToSocketAddrs, IpAddr};
+use std::fs::File;
+use std::path::PathBuf;
+
+/// Listens on 127.0.0.1:8000 and opens `source_peers` sockets from the
+/// computations we're examining (one socket for every worker on the
+/// examined computation).
+///
+/// The sockets are returned in nonblocking mode.
+pub fn open_sockets(ip_addr: IpAddr, port: u16, source_peers: usize) -> Result<Vec<TcpStream>, ConnectError> {
+    let socket_addr = (ip_addr, port).to_socket_addrs()?
+        .next().ok_or(ConnectError::Other("Invalid listening address".to_string()))?;
+    let listener = TcpListener::bind(socket_addr)?;
+    Ok((0..source_peers).map(|_| {
+        let socket = listener.incoming().next().expect("Socket unexpectedly unavailable");
+        if let Ok(ref s) = &socket {
+            s.set_nonblocking(true)?;
+        }
+        socket
+    }).collect::<Result<Vec<_>, _>>()?)
+}
+
+/// Types of Read created by `make_replayers`
+pub enum TcpStreamOrFile {
+    /// a TCP-backed online reader
+    Tcp(TcpStream),
+    /// a file-backed offline reader
+    File(File),
+}
+
+impl std::io::Read for TcpStreamOrFile {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        match self {
+            TcpStreamOrFile::Tcp(x) => x.read(buf),
+            TcpStreamOrFile::File(x) => x.read(buf),
+        }
+    }
+}
+
+/// Source of binary data for `make_replayers`.
+pub enum ReplaySource {
+    Tcp(Arc<Mutex<Vec<Option<TcpStream>>>>),
```

Comment:
Why do these need to be wrapped into an `Option`?

Reply:
Each worker needs to `.take()` the streams assigned to it. Can you see a way to do this without the `Option`s?

Reply:
Ah, that makes sense. Could you then make `open_sockets` return a `Vec<Option<TcpStream>>`? Otherwise I have to re-wrap the sockets manually after calling `open_sockets`.